File: response_test.go
package main

import (
	"reflect"
	"testing"
)

func TestResponse_newResponse(t *testing.T) {
	actual := newResponse()
	expected := new(response)
	expected.Items = []responseItem{}

	if !reflect.DeepEqual(actual, expected) {
		t.Errorf("expected %v to eq %v", actual, expected)
	}
}

func TestResponse_addItem(t *testing.T) {
	r := newResponse()

	actual := r.addItem(&responseItem{Title: "title-foo"}).Items
	expected := []responseItem{responseItem{Title: "title-foo"}}

	if !reflect.DeepEqual(actual, expected) {
		t.Errorf("expected %v to eq %v", actual, expected)
	}
}

func TestResponse_toXML(t *testing.T) {
	r := newResponse()
	item := responseItem{
		Valid:    true,
		UID:      "f000-uid",
		Title:    "title-foo",
		Subtitle: "Subtitle foo.",
		Arg:      "arg-foo",
		Icon:     "./icons/title-foo.png",
		Unicode:  "f000-unicode",
	}
	r.addItem(&item)

	actual, err := r.toXML()
	if err != nil {
		t.Error("failed to convert to XML")
		return
	}

	expected := `<?xml version="1.0" encoding="UTF-8"?>
<items><item valid="true" arg="arg-foo" uid="f000-uid" unicode="f000-unicode"><title>title-foo</title><subtitle>Subtitle foo.</subtitle><icon>./icons/title-foo.png</icon></item></items>`
	if actual != expected {
		t.Errorf("expected %v to eq %v", actual, expected)
	}
}

func TestResponse_toXML_Blank(t *testing.T) {
	r := newResponse()

	actual, err := r.toXML()
	if err != nil {
		t.Error("failed to convert to XML")
		return
	}

	expected := `<?xml version="1.0" encoding="UTF-8"?>
<items></items>`
	if actual != expected {
		t.Errorf("expected %v to eq %v", actual, expected)
	}
}

File: src/alfred/response.test.ts
import { toJson, TResponse } from "./response";

const response: TResponse = {
  items: [
    {
      title: "500px",
      subtitle: "Paste class name: fa-500px",
      arg: "f26e",
      icon: {
        type: "fileicon",
        path: "./icons/500px.png",
      },
    },
    {
      title: "calendar-plus",
      subtitle: "Paste class name: fa-calendar-plus",
      arg: "f217",
      icon: {
        type: "fileicon",
        path: "./icons/calendar-plus.png",
      },
    },
  ],
};

test("toJson", () => {
  const actual = toJson(response);
  const expected =
    '{"items":[{"title":"500px","subtitle":"Paste class name: fa-500px","arg":"f26e","icon":{"type":"fileicon","path":"./icons/500px.png"}},{"title":"calendar-plus","subtitle":"Paste class name: fa-calendar-plus","arg":"f217","icon":{"type":"fileicon","path":"./icons/calendar-plus.png"}}]}';
  expect(actual).toBe(expected);
});
